### DISCUSSION TITLE
Fail installation if a package in the store has been modified?

### DISCUSSION THREAD

**Discussion by zkochan** [pnpm/pnpm#16189] [Ideas]
People sometimes modify packages in the node_modules folder. When they do it, the package is modified in the store as well.

During installation, if a package in the store has been modified, we redownload it from the registry and print a message: `WARN Refetching /home/zkochan/.pnpm-store/1/registry.npmjs.org/pnpm-registry-mock/0.7.0 to store, as it was modified`

Should we instead fail the installation? Users will be able to run installation with the `--force` parameter if they are sure they are OK rewriting their changes.

With the current behavior I am afraid someone will be mad at pnpm for loss of his/her work, as in #724

**Comment by memotux** [comment]
Measured this on both review versions, since the thread is empty and the
surrounding issues have moved on. Everything below is from runs I did today.

Setup: pnpm 12.5.1 and 11.27.0, Ubuntu guest (arm64, native ext4, kernel
7.0.0-28-generic), `is-number@7.0.0`, warm per-run store. The store's content
file for `index.js` is overwritten in place, with the sha256 recorded before and
after.

```bash
pnpm install                        # links is-number@7.0.0 from a warm store
printf 'TAMPERED' > "$STORE_FILE"   # in-place write to the store copy
pnpm install                        # exit 0, "Already up to date"
pnpm install --force                # exit 0, project copy still TAMPERED
pnpm store status -y                # exit 1, ERR_PNPM_MODIFIED_DEPENDENCY
```

Observed, identical on 12.5.1 and 11.27.0:

| Step | Result |
|---|---|
| `install` in the project that already holds the package | exit 0, `Already up to date`, store file and project `index.js` both still the modified bytes |
| `install` in a fresh project on the same store | refetches (`Progress: ... downloaded 1`), store object restored (new inode, pristine sha256), fresh project gets correct content |
| `install --force` in the project that already holds it | exit 0, project `index.js` still the modified bytes |
| `store status -y` | exit 1, `ERR_PNPM_MODIFIED_DEPENDENCY`, and the `help:` line pointing at `pnpm install --force` |

Against the expectation in that help text: `--force` did not repair the project
that consumed the modified bytes. The store object is restored by the next
install that links it fresh, so `--force` can appear to have worked when the
store is already pristine. I could not demonstrate that `--force` itself repairs
a modified store object, because on Linux the store had already been restored
before `--force` ran.

Scope limits: I did not test the metadata case from #14173 (zeroing
`package.json` fails loudly with `ERR_PNPM_JSON_PARSE`, which is a different mode
from the silent one above), and my macOS runs are not usable here, because the
copy import there links a different store file and `store status` reported the
store untouched while a fresh project still picked up the modified bytes. Only
the Linux runs are reported.

What this leaves for the decision: #14173 is closed as a duplicate of #919, and
#919 was closed by dropping the hint (#15359) rather than making it true, with
#15162 and #15192 in flight for `--force` on repeat installs. The store object
does get repaired on the next fresh link, so if install should fail or warn, the
case that matters is the project that already consumed the modified bytes,
because there nothing surfaces at all today.

Options as I read them: fail the install when a store object does not match its
recorded checksum, with an explicit override (the original proposal here); warn
once and continue; or keep detection on demand via `store status`. My suggestion
is that any of them should cover the already-installed project, since the
fresh-link path already behaves sensibly and the silent case is the one the old
help text misled users about. Making that check means verifying store objects
that are not read at all today, so its cost is worth stating when the choice is
made.

Happy to share the harness and the full output for both versions.
